### DISCUSSION TITLE
WoX plugin migration

### DISCUSSION THREAD

**Discussion by ruslanmogilevskiy** [Flow-Launcher/Flow.Launcher#1033] [Q&A]
I`m migrating a plugin from WoX code base but faced with some troubles in code compatibility:
1) I`m using PluginJsonStorage<> to save and load the plugin settings. How to access it in Flow? 
2) In WoX, to access PluginJsonStorage I was referencing WoX.Plugin.Infrastructure assembly. How to reference  Flow.Launcher.Infrastructure? Is there a nuget package?
3) To test my plugin in WoX or quickly install other plugins as the *.wox files I could just drag & drop a .wox file to the WoX`s query window and WoX suggests to install the plugin. The same behaviour does not work in Flow. Is this feature was removed? Why?

**Comment by taooceros** [comment]
1. You can use the loadjsonstorage in the public api as an instead of plugindtorage.
2. We stop requiring Plugins to reference the infrastructure so everything a plugin needs is in Flow.launcher.plugin
3. I personally don't know about that feature (it is removed before I join flow launcher). But if it is very convenient, we can add it back in the future. We do have the way to install a zip file from url so it should be easy.

**Comment by jjw24** [comment]
1. You can also have a look at the existing plugins that shipped with flow to see how it's used as an example.
3. It's easier to test if you just set your project's build output folder to the userdata's plugin folder temporarily so everytime you build new changes will be updated, rather than zipping up the files then try to drag and drop it in manually.

**Reply by ruslanmogilevskiy** [reply]
@jjw24,

Thanks for your help.
1) I looked at other plugins but didn`t find how they store/retrieve their own settings. What I saw, is that the plugin (de)serializes the settings into a local file by itself, which is not useful at all.
2) IMO, your suggestion (to build locally) isn`t handy because that preparations (creating a ZIP file) could be automated and run upon the build w/o manual work. And, what IMO is super useful, is that you can get/download any plugin`s zipped file (.wox for WoX) and just drag it over the FlowLauncher. You won`t need to know where plugins are situated, how to install it manually, and so on. I used it in WoX all the time, and it was cool UX.

**Reply by Garulf** [reply]
You can type `pm install http://<URL to plugin zip file>` and Flow will install it for you. No need to download manually and drag & drop.

**Reply by ruslanmogilevskiy** [reply]
Some plugins are not published as a plugin file, only sources. Some plugins could be shared on forums, for instance. Also, for local development, it is useful to be able to quickly install the plugin in this way.

What was the reason for removing such a handy WoX's feature and then discussing how we can replicate it manually?

**Reply by jjw24** [reply]
> I looked at other plugins but didn`t find how they store/retrieve their own settings. What I saw, is that the plugin (de)serializes the settings into a local file by itself, which is not useful at all.

If I remember correctly you just need to init the settings file with PluginJsonStorage, which yes saves it to the standard settings location in the userdata location. I'm afk so will point you to it when I get back. Unless it's not what you are after, are you wanting to save to a different location or format?

**Reply by jjw24** [reply]
>IMO, your suggestion (to build locally) isnt handy because that preparations (creating a ZIP file) could be automated and run upon the build w/o manual work. And, what IMO is super useful, is that you can get/download any plugins zipped file (.wox for WoX) and just drag it over the FlowLauncher. You won`t need to know where plugins are situated, how to install it manually, and so on. I used it in WoX all the time, and it was cool UX.

I don't mean you need to zip it up, I mean when you are building your plugin, you can set your build output location to userdata>plugins location, which will output the contents of your build there and when you restart flow new changes will be loaded again, (for c# plugins you need to restart flow, non-c# ones it's hot reload which means new changes reflected instantly). 

So no zipping up is involved. The method @Garulf mentioned is where you want to download the hosted plugin from somewhere else you can use that method to directly download them.

I havent fully used the way Wox did their where you drag a .wox file into the query box, but I could imagine you need to first convert/zip it up in order to drag it in. If you are talking about downloading the .wox file from their website then drag in, then the method mentioned by @Garulf is even faster where you just paste in the link, not needing to download. If you are mentioning that you are building a plugin locally, then shouldn't need to convert it to zip or .wox, you could just do my method where you output the build contents into the plugin folder.

The drag drop feature is missing because our plugin hosting system is different to wox and we rewrote the plugin that handles the plugin management system and the feature was not considered. Obviously happy to add a better solution but if the above I mentioned still not working, let's keep the conversation going, must be some context we missed.

**Reply by jjw24** [reply]
>Some plugins are not published as a plugin file, only sources. Some plugins could be shared on forums, for instance. Also, for local development, it is useful to be able to quickly install the plugin in this way.

Try this as an example to install the Colours plugin:

`pm install https://github.com/Flow-Launcher/Flow.Launcher.Plugin.Color/releases/download/v2.0.1/Color-2.0.1.zip`

**Reply by ruslanmogilevskiy** [reply]
@jjw24, my point is that in WoX it was VERY easy to install the plugin, you just d&d it over there, WoX is asking you whether to install it and then auto-restarts itself, voila, the plugin is there. No need to enter http://.. commands or do some manual job. 
Anyway, thanks for helping with the state loading, I'll check it. My goal is just to save/load a simple state to a standard place/storage/file that Flow provides. In WoX it was just using the PluginJsonStorage to do so without thinking how and when it's stored and I was looking for smth simple like that but in Flow.

**Reply by jjw24** [reply]
> to install it and then auto-restarts itself, voila, the plugin is there. No need to enter http://.. commands or do some manual job.

hmm I must be missing something because it's still similar, just instead of downloading the file, drag and drop into the query window, you use `pm install` paste the url path to the zip file and it will prompt, install and restart for you.

> My goal is just to save/load a simple state to a standard place/storage/file that Flow provides. In WoX it was just using the PluginJsonStorage to do so without thinking how and when it's stored and I was looking for smth simple like that but in Flow.

This is what you need to do (assuming you are developing C# plugin):
Loads the settings file
https://github.com/Flow-Launcher/Flow.Launcher/blob/16e032ea551d788c5a1d9bfad84d00bc12d1db87/Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs#L27

For set up in Flow's settings window:
https://github.com/Flow-Launcher/Flow.Launcher/blob/16e032ea551d788c5a1d9bfad84d00bc12d1db87/Plugins/Flow.Launcher.Plugin.BrowserBookmark/Main.cs#L94-L97

Your settings class is like this for example:
https://github.com/Flow-Launcher/Flow.Launcher/blob/16e032ea551d788c5a1d9bfad84d00bc12d1db87/Plugins/Flow.Launcher.Plugin.BrowserBookmark/Models/Settings.cs#L7-L15

And that's it. Flow will handle the saving of any new changes to your settings automatically. The settings.json file is saved in UserData > Settings > Plugins.
